### PR TITLE
lz4 1.9.4

### PR DESCRIPTION
**Pull request by jcfj**
(Seems like lz4's contrib directory is not protected by semver. Still, I'd prefer to use the files from upstream as is.)

### CODE REVIEW

**Comment by jcfj** [comment]
I wanted to make this two separate PRs, but… 

>  Pull request creation failed. Validation failed: You can't perform that action at this time. 

I guess I'm a spammer.

**Comment by neheb** [comment]
```
2022-08-16T06:35:17.5310008Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(91): error C2491: 'cJSON_GetErrorPtr': definition of dllimport function not allowed
2022-08-16T06:35:17.5310929Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(95): error C2491: 'cJSON_GetStringValue': definition of dllimport function not allowed
2022-08-16T06:35:17.5311764Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(103): error C2491: 'cJSON_GetNumberValue': definition of dllimport function not allowed
2022-08-16T06:35:17.5313847Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(118): error C2491: 'cJSON_Version': definition of dllimport function not allowed
2022-08-16T06:35:17.5314688Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(197): error C2491: 'cJSON_InitHooks': definition of dllimport function not allowed
2022-08-16T06:35:17.5315505Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(235): error C2491: 'cJSON_Delete': definition of dllimport function not allowed
2022-08-16T06:35:17.5316342Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(361): error C2491: 'cJSON_SetNumberHelper': definition of dllimport function not allowed
2022-08-16T06:35:17.5317092Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(374): error C2491: 'cJSON_SetValuestring': definition of dllimport function not allowed
2022-08-16T06:35:17.5317914Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(994): error C2491: 'cJSON_ParseWithOpts': definition of dllimport function not allowed
2022-08-16T06:35:17.5318792Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1013): error C2491: 'cJSON_ParseWithLengthOpts': definition of dllimport function not allowed
2022-08-16T06:35:17.5319630Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1085): error C2491: 'cJSON_Parse': definition of dllimport function not allowed
2022-08-16T06:35:17.5320447Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1090): error C2491: 'cJSON_ParseWithLength': definition of dllimport function not allowed
2022-08-16T06:35:17.5321257Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1157): error C2491: 'cJSON_Print': definition of dllimport function not allowed
2022-08-16T06:35:17.5322028Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1161): error C2491: 'cJSON_PrintUnformatted': definition of dllimport function not allowed
2022-08-16T06:35:17.5322834Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1166): error C2491: 'cJSON_PrintBuffered': definition of dllimport function not allowed
2022-08-16T06:35:17.5323645Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1196): error C2491: 'cJSON_PrintPreallocated': definition of dllimport function not allowed
2022-08-16T06:35:17.5324762Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1686): error C2491: 'cJSON_GetArraySize': definition of dllimport function not allowed
2022-08-16T06:35:17.5325851Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1722): error C2491: 'cJSON_GetArrayItem': definition of dllimport function not allowed
2022-08-16T06:35:17.5326604Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1764): error C2491: 'cJSON_GetObjectItem': definition of dllimport function not allowed
2022-08-16T06:35:17.5336156Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1770): error C2491: 'cJSON_GetObjectItemCaseSensitive': definition of dllimport function not allowed
2022-08-16T06:35:17.5336945Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1775): error C2491: 'cJSON_HasObjectItem': definition of dllimport function not allowed
2022-08-16T06:35:17.5337804Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1833): error C2491: 'cJSON_AddItemToArray': definition of dllimport function not allowed
2022-08-16T06:35:17.5338505Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1893): error C2491: 'cJSON_AddItemToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5339195Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1899): error C2491: 'cJSON_AddItemToObjectCS': definition of dllimport function not allowed
2022-08-16T06:35:17.5340760Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1904): error C2491: 'cJSON_AddItemReferenceToArray': definition of dllimport function not allowed
2022-08-16T06:35:17.5341760Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1913): error C2491: 'cJSON_AddItemReferenceToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5342542Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1924): error C2491: 'cJSON_AddNullToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5343362Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1935): error C2491: 'cJSON_AddTrueToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5344316Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1946): error C2491: 'cJSON_AddFalseToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5344989Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1960): error C2491: 'cJSON_AddBoolToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5345676Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1973): error C2491: 'cJSON_AddNumberToObject': definition of dllimport function not allowed
2022-08-16T06:35:17.5346341Z ../subprojects/librdkafka-1.9.2/src/cJSON.c(1987): error C2491: 'cJSON_AddStringToObject': definition of dllimport function not allowed
```

**Comment by tristan957** [comment]
Keep an eye on https://github.com/lz4/lz4/pull/1139

**Comment by neheb** [comment]
```
ERROR: File 'subprojects\\lz4-1.9.4\\contrib/meson/meson\\lib\\lz4.lib' could not be found
```

dllexport strikes again.

**Comment by eli-schwartz** [comment]
I would prefer not to add 350 lines worth of inclined licenses that aren't there in the upstream file.

**Comment by neheb** [comment]
SPDX is probably better.

**Comment by jcfj** [comment]
> I would prefer not to add 350 lines worth of inlined licenses that aren't there in the upstream file.

Those used to be in separate files, [but this check got added](https://github.com/mesonbuild/wrapdb/runs/7851652691?check_suite_focus=true). (i.e. I didn't add them, just moved them. Would you prefer to have them gone?)

> dllexport strikes again.

Sorry, I have no idea what's going on with that.

File: ci_config.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "arduinocore-avr",
     "google-benchmark", "google-woff2", "graphite2",
     "libgpiod",
-    "lz4", "mocklibc", "openal-soft",
```
Comment (neheb) [inline ci_config.json:26]: Should be added to Windows

**Comment by neheb** [comment]
Aside from duplicating and fixing the meson.build files, I wonder if it's possible to override the upstream library definition to one that provides a vs_module_defs

**Comment by tristan957** [comment]
Why not just submit a patch upstream?

**Comment by tristan957** [comment]
Where I am at with the upstream PR: Working on getting tests going. Tests that are defined inline in Makefiles are the worst thing to ever exist :(. I would recommend looking at that PR and playing around with it. Please let me know if you think something is missing because we only get 1 shot at making sure everything is good to go until we have to wait another cycle for an lz4 release!

**Comment by tristan957** [comment]
What versions of Meson do you all want lz4 to support? Kinda jumping through a few hoops in the upstream PR.

File: ci_config.json [hunk 2]
```diff
@@ -14,7 +14,7 @@
   "broken_windows": [
     "arduinocore-avr",
     "google-woff2",
-    "libgpiod",
+    "libgpiod", "lz4",
```
Comment (neheb) [inline ci_config.json:17]: shouldn't be here. Rather a separate lz4 entry should be created and build_on windows should be false. See https://github.com/mesonbuild/wrapdb/blob/master/ci_config.json#L413

Comment (eli-schwartz) [reply ci_config.json]: The upstream software is intended to work on Windows, but instead it is broken -- which means that broken_windows is correct, and build_on is incorrect...

Comment (jcfj) [reply ci_config.json]: `build_on` (+ rebase on master) seem to have worked.

@eli-schwartz What would I need to do to get past the [CI](https://github.com/jcaesar/wrapdb/runs/8135452483?check_suite_focus=true) with that?

Comment (eli-schwartz) [reply ci_config.json]: Make it actually work!!!!

Good grief.

**Comment by eli-schwartz** [comment]
> I would prefer not to add 350 lines worth of inlined licenses that aren't there in the upstream file.

Ignored? Why?

Also this is broken and breaks downstream consumers. This PR needs to be reverted.

I'm strongly tempted to remove ALL support for allowing or permitting wraps that don't work on Windows, simply because people keep on ***cheating*** and disabling CI because "there's an option to disable CI'.

**Comment by jcaesar** [comment]
> Ignored? Why?

Because you ignored

> Those used to be in separate files, [but this check got added](https://github.com/mesonbuild/wrapdb/runs/7851652691?check_suite_focus=true). (i.e. I didn't add them, just moved them. Would you prefer to have them gone?)

File: ci_config.json (hunk 2 above)
Comment (jcaesar) [reply ci_config.json]: > Good grief.

Good joke. I haven't done any development on windows in more than a decade and I have no idea whether that's a problem with the CI (I don't really know what your CI does, after all…) which is why I was asking.

Thankfully, somebody else took care of it. I suppose I'll stop submitting PRs..

**Comment by tristan957** [comment]
Windows support should be working upstream now

**Comment by eli-schwartz** [comment]
> Those used to be in separate files, [but this check got added](https://github.com/mesonbuild/wrapdb/runs/7851652691?check_suite_focus=true). (i.e. I didn't add them, just moved them. Would you prefer to have them gone?)

Every other project with a license checked into the repo just calls it `LICENSE.build` :/

File: ci_config.json (hunk 2 above)
Comment (eli-schwartz) [reply ci_config.json]: What mostly annoys me about this is not that you tried to use an option someone told you to use. What annoys me is that someone told you to use it.

That being said, if the build files come from upstream, and "The upstream software is intended to work on Windows, but instead it is broken"... my instinctive assumption would be the first thing to try is to report an upstream bug.

> and I have no idea whether that's a problem with the CI (I don't really know what your CI does, after all…) which is why I was asking

And I answered that the problem with the CI indicates the software is broken.

But @neheb thinks that software being broken means that PRs should be merged anyway, so idk. Clearly there's a difference of opinion here. Maybe I'm a fool after all.

Comment (eli-schwartz) [reply ci_config.json]: > I have no idea whether that's a problem with the CI

It's a legitimate problem that the CI *allows* people to do things that aren't good per policy, and for that I apologize. We can, in fact, do better.

I updated the CI harness: I have no idea whether that's a problem with the CI

Now, the CI will actually tell you if something is okay or not.

**Pull request merged**